crates/node/src/lib.rs

```diff
@@ -77,67 +56,85 @@ impl From<GlobEntry> for tailwindcss_oxide::GlobEntry {
 
 impl From<tailwindcss_oxide::GlobEntry> for GlobEntry {
   fn from(glob: tailwindcss_oxide::GlobEntry) -> Self {
-    GlobEntry {
+    Self {
       base: glob.base,
       pattern: glob.pattern,
     }
   }
 }
 
+impl From<AutoContent> for tailwindcss_oxide::scanner::auto_content::AutoContent {
+  fn from(auto_content: AutoContent) -> Self {
+    Self::new(auto_content.base.into())
+  }
+}
+
+// ---
+
 #[derive(Debug, Clone)]
 #[napi(object)]
-pub struct ScanOptions {
-  /// Base path to start scanning from
-  pub base: Option<String>,
+pub struct ScannerOptions {
+  /// Auto content configuration
+  pub auto_content: Option<AutoContent>,
+
   /// Glob sources
   pub sources: Option<Vec<GlobEntry>>,
 }
 
+#[derive(Debug, Clone)]
 #[napi]
-pub fn clear_cache() {
-  tailwindcss_oxide::clear_cache();
+pub struct Scanner {
+  scanner: tailwindcss_oxide::Scanner,
 }
 
 #[napi]
-pub fn scan_dir(args: ScanOptions) -> ScanResult {
-  let result = tailwindcss_oxide::scan_dir(tailwindcss_oxide::ScanOptions {
-    base: args.base.clone(),
-    sources: args
-      .sources
-      .clone()
-      .unwrap_or_default()
-      .into_iter()
-      .map(Into::into)
-      .collect(),
-  });
-
-  ScanResult {
-    // Private
-    base: args.base,
-    sources: args.sources.unwrap_or_default(),
-
-    // Public
-    files: result.files,
-    candidates: result.candidates,
-    globs: result.globs.into_iter().map(Into::into).collect(),
+impl Scanner {
+  #[napi(constructor)]
+  pub fn new(opts: ScannerOptions) -> Self {
+    Self {
+      scanner: tailwindcss_oxide::Scanner::new(
+        opts.auto_content.map(Into::into),
+        opts
+          .sources
+          .map(|x| x.into_iter().map(Into::into).collect()),
+      ),
+    }
   }
-}
 
-#[derive(Debug)]
-#[napi]
-pub enum IO {
-  Sequential = 0b0001,
-  Parallel = 0b0010,
-}
+  #[napi]
+  pub fn scan_files(&mut self, input: Vec<ChangedContent>) -> bool {
+    let candidate_count_before = self.scanner.total_candidates();
 
-#[derive(Debug)]
-#[napi]
-pub enum Parsing {
-  Sequential = 0b0100,
-  Parallel = 0b1000,
-}
+    // Scan new content
+    self
+      .scanner
+      .scan_content(input.into_iter().map(Into::into).collect());
 
-#[napi]
-pub fn scan_files(input: Vec<ChangedContent>, strategy: u8) -> Vec<String> {
-  tailwindcss_oxide::scan_files(input.into_iter().map(Into::into).collect(), strategy)
+    let candidate_count_after = self.scanner.total_candidates();
+
+    // Return whether we scanned new candidates or not. We can do this because:
+    // 1. It's a set, so we can't have duplicates
+    // 2. We're only adding new candidates, not removing them
+    candidate_count_before != candidate_count_after
+  }
+
+  #[napi]
+  pub fn get_candidates(&self) -> Vec<String> {
+    self.scanner.get_candidates()
+  }
```

Comment:
What is the behavior if a vec is returned here and exported to Node via napi? My current thinking is that this needs to copy the whole array every time we call getCandidates which might be quite a bit of I/O for large projects.

Maybe alternatively we could only return the new candidates (so moving some of the bookeeping into rust by keeping it in a set that we can discriminate between new and old).

I understand that it is annoying to collect the results in JS land in another data structure but I imagine that we would otherwise call `getCandidates()` quite often in dev mode so maybe this is a reasonable optimization?

Reply:
Yeah that's right. Thinking about it, I think we even track all known candidates in tailwind's core already. So this can definitely be simplified.

Reply:
If we do that I think it needs a different name too — maybe something like "newCandidates" or w/e.

On the flip side — even with this stuff being copied — I'd see if we can measure the overhead to see if it's significant or not.